### ISSUE TITLE
Reproducing result form SplineCNN result using PyG

### ISSUE THREAD

**Issue by adhikarirsr** [pyg-team/pytorch_geometric#1242]
I am trying to replicate SplineCNN's MNIST result. I only got about 70% accuracy( as opposed to ~99% reported in the paper). Here's my attempt:

```
#Importing modules:
import torch
import torch.nn.functional as F
import torch_geometric.transforms as T
from torch_geometric.data import Data
from torch_geometric.utils import grid,normalized_cut
from torch_geometric.nn import (SplineConv, graclus, max_pool, max_pool_x,
                                global_mean_pool)
import numpy as np
```

**1. Loading MNIST data:( I am using spektral to load MNIST data as I started learning GNNs using spektral and the mnist load module is very flexible)** 

```
!pip install spektral
from spektral.datasets import mnist
X_train, y_train, X_val, y_val, X_test, y_test, Aspek = mnist.load_data()
#concatenating train and val data 
X_train = np.concatenate((X_train,X_val),axis=0)
y_train = np.concatenate((y_train,y_val),axis=0)
```
**2. Creating dataset:**

```
#creating grid graph
pos1 = grid(28,28)
```

```
#creating MNIST train data 
from torch_geometric.data import InMemoryDataset

class MNISTDataset(InMemoryDataset):
    def __init__(self, root, transform=None, pre_transform=None):
        super(MNISTDataset, self).__init__(root, transform, pre_transform)
        self.data, self.slices = torch.load(self.processed_paths[0])

    @property
    def raw_file_names(self):
        return []

    @property
    def processed_file_names(self):
        return ['/tmp/mdata.pt']

    def download(self):
        pass

    def process(self):

        data_list= []
        for i in range(0, len(X_train)):
            data = Data(x=torch.tensor(X_train[i],dtype=torch.float).view(784,1)
            ,edge_index=pos1[0],y=torch.tensor(y_train[i]).view(1).long(),
            pos=pos1[1])

            data_list.append(data)

        data, slices = self.collate(data_list)
        torch.save((data, slices), self.processed_paths[0])

```

```
#creating MNIST test data 

class MNISTtestDataset(InMemoryDataset):
    def __init__(self, root, transform=None, pre_transform=None):
        super(MNISTtestDataset, self).__init__(root, transform, pre_transform)
        self.data, self.slices = torch.load(self.processed_paths[0])

    @property
    def raw_file_names(self):
        return []

    @property
    def processed_file_names(self):
        return ['/tmp/mdata1.pt']

    def download(self):
        pass

    def process(self):

        data_list= []
        for i in range(0, len(X_test)):
            data = Data(x=torch.tensor(X_test[i],dtype=torch.float).view(784,1)
            ,edge_index=pos1[0],y=torch.tensor(y_test[i]).view(1).long(),
            pos=pos1[1])

            data_list.append(data)

        data, slices = self.collate(data_list)
        torch.save((data, slices), self.processed_paths[0])
```
`transform = T.Cartesian(cat=False)`
```
datasetrain = MNISTDataset(root='/tmp/Mdtrain', transform=transform)
datasetest = MNISTtestDataset(root='/tmp/Mdtest', transform=transform)
```

```
# Creating dataloader for training and testing
from torch_geometric.data import DataLoader
train_loader = DataLoader(datasetrain, batch_size=64, shuffle=True)
test_loader = DataLoader(datasetest, batch_size=64)
```

**3. Model**
```
d = datasetrain
def normalized_cut_2d(edge_index, pos):
    row, col = edge_index
    edge_attr = torch.norm(pos[row] - pos[col], p=2, dim=1)
    return normalized_cut(edge_index, edge_attr, num_nodes=pos.size(0))

class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = SplineConv(d.num_features, 32, dim=2, kernel_size=5)
        self.conv2 = SplineConv(32, 64, dim=2, kernel_size=5)
        self.fc1 = torch.nn.Linear(64, 512)
        self.fc2 = torch.nn.Linear(512, d.num_classes)

    def forward(self, data):
        data.x = F.elu(self.conv1(data.x, data.edge_index, data.edge_attr))
        weight = normalized_cut_2d(data.edge_index, data.pos)
        cluster = graclus(data.edge_index, weight, data.x.size(0))
        data.edge_attr = None
        data = max_pool(cluster, data, transform=transform)

        data.x = F.elu(self.conv2(data.x, data.edge_index, data.edge_attr))
        weight = normalized_cut_2d(data.edge_index, data.pos)
        cluster = graclus(data.edge_index, weight, data.x.size(0))
        x, batch = max_pool_x(cluster, data.x, data.batch)

        x = global_mean_pool(x, batch)
        x = F.elu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        return F.log_softmax(self.fc2(x), dim=1)
```
**4.Training and Testing**
```
#Training module
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = Net().to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=0.01, weight_decay=5e-4)

def train(epoch):
    model.train()

    for data in train_loader:
        data = data.to(device)
        optimizer.zero_grad()
        F.nll_loss(model(data), data.y).backward()
        optimizer.step()

def test():
    model.eval()
    correct = 0

    for data in test_loader:
        data = data.to(device)
        pred = model(data).max(1)[1]
        correct += pred.eq(data.y).sum().item()
    return correct / len(datasetest)

for epoch in range(1, 50):
    train(epoch)
    test_acc = test()
    print('Epoch: {:02d}, Test: {:.4f}'.format(epoch, test_acc))
```

What am I missing here? Any guidance would be helpful.

**Comment by arshishir** [comment]
@rusty1s Could you please let me know what am I doing wrong here?

**Comment by rusty1s** [comment]
I will try to rerun it by myself.

**Comment by rusty1s** [comment]
Only looking at your code, there are a few differences to the architecture we use:
1. We use a 2-hop neighboring grid with kernel size 5 (which would be nearly identical to stacking 2 convs with kernel_size=3x3)
2. We do not make use of global mean pooling before the final classification, and instead flatten the input representation.

**Comment by arshishir** [comment]
@rusty1s  Thanks for replying.

1. Do you think my dataset generation is correct? I understand that for SplineCNN position is necessary as it is based on spatial graph conv. So, I created my dataset to reflect that. 

2. I am following this : `SConv((5, 5),1,32)→ MaxP(4) → SConv((5, 5),32,64) → MaxP(4) →FC(512) → FC(10)` from your paper. I thought the code above reflected that except for the global max pool. What would the code look like for the arch you have on your paper?

3. I understand that the SplineConv that I am using does not have a huge receptive field. However, I thought the network above should be good enough to give me over 95% accuracy on MNIST.

**Comment by rusty1s** [comment]
I'm not sure your dataset generation is correct. Spektral may use a different ordering? I always used the dataset provided by `torchvision`.

I prepared two simple scripts that should get you going. One that uses `graclus` and `global_mean_pool` (which reaches around 92% after 5 epochs already) and one that uses the flattening approach without graclus (which reaches around 97% after 1 epoch). You can find them [here](https://gist.github.com/rusty1s/74b1efe973482e070c50f650ece82b70) and [here](https://gist.github.com/rusty1s/866ca8b830f22aa34923a0f8164f6e64).

**Comment by arshishir** [comment]
Thanks @rusty1s. I will go through your codes and compare them with mine to see what I am missing. Thanks again for your help.
